src/scene/batching.js

```diff
@@ -0,0 +1,837 @@
+    };
+
+    SkinBatchInstance.prototype = {
+
+        updateMatrices: function () {
+        },
+
+        updateMatrixPalette: function () {
+            var pe;
+            var mp = this.matrixPalette;
+            var base;
+
+            for (var i = this.bones.length - 1; i >= 0; i--) {
+                pe = this.bones[i].getWorldTransform().data;
+
+                // Copy the matrix into the palette, ready to be sent to the vertex shader
+                base = i * 16;
+                mp[base] = pe[0];
+                mp[base + 1] = pe[1];
+                mp[base + 2] = pe[2];
+                mp[base + 3] = pe[3];
+                mp[base + 4] = pe[4];
+                mp[base + 5] = pe[5];
+                mp[base + 6] = pe[6];
+                mp[base + 7] = pe[7];
+                mp[base + 8] = pe[8];
+                mp[base + 9] = pe[9];
+                mp[base + 10] = pe[10];
+                mp[base + 11] = pe[11];
+                mp[base + 12] = pe[12];
+                mp[base + 13] = pe[13];
+                mp[base + 14] = pe[14];
+                mp[base + 15] = pe[15];
+            }
+
+            if (this.device.supportsBoneTextures) {
+                this.boneTexture.lock();
+                this.boneTexture.unlock();
+            }
+        }
+    };
+
+    /**
+     * @name pc.BatchManager
+     * @class Glues many mesh instances into a single one for better performance.
+     */
+    var BatchManager = function (device, root, scene) {
+        this.device = device;
+        this.rootNode = root;
+        this.scene = scene;
+        this._init = false;
+
+        this._batchGroups = {};
+        this._batchGroupNameToId = {};
+        this._batchGroupCounter = 0;
+
+        // #ifdef PROFILER
+        this._stats = {
+            createTime: 0,
+            updateLastFrameTime: 0
+        };
+        // #endif
+    };
+
+    /**
+     * @private
+     * @function
+     * @name pc.BatchManager#addGroup
+     * @description Adds new global batching group.
+     * @param {String} name Custom name
+     * @param {Boolean} dynamic Is this batching group dynamic? Will these objects move/rotate/scale after being batched?
+     * @param {Number} maxAabbSize Maximum size of any dimension of a bounding box around batched objects.
+     * BatchManager.prepare() will split objects into local groups based on this size.
+     * @returns {pc.BatchGroup} Group object.
+     */
+    BatchManager.prototype.addGroup = function(name, dynamic, maxAabbSize) {
+        if (this._batchGroupNameToId[name]) {
+            // #ifdef DEBUG
+            console.error("batch group " + name + " already exists");
+            // #endif
+            return;
+        }
+        var id = this._batchGroupCounter;
+        var group;
+        this._batchGroups[id] = group = new pc.BatchGroup(id, name, dynamic, maxAabbSize);
+        this._batchGroupNameToId[name] = id;
+        this._batchGroupCounter++;
+
+        this._batchList = [];
```

Comment:
This should be initialized in constructor too, otherwise it is undefined if no group is added